File: pkg/types/task.go
package types

import (
	"encoding/json"
	"time"
)

// WorkflowMode identifies a Hecate-owned, bounded task execution contract.
// It is intentionally distinct from Cairnline coordination: a workflow mode
// changes how Hecate executes and records a task, not how work is planned or
// assigned across a project.
type WorkflowMode string

const (
	// WorkflowModeQA is Hecate's report-only QA runbook. The runtime owns its
	// read-only tool posture and report artifacts; callers cannot turn those
	// restrictions into a broader agent workflow with request fields.
	WorkflowModeQA WorkflowMode = "qa"
)

type Task struct {
	ID     string
	Title  string
	Prompt string
	// ProjectID links the task to an operator project when it was
	// created from a project-scoped surface (manual Tasks, Hecate Chat,
	// or project assignments). Empty means the task is unscoped.
	ProjectID string
	// WorkItemID and AssignmentID link a task to the project work records
	// that created it. Empty means the task was project-scoped without a
	// project assignment.
	WorkItemID   string
	AssignmentID string
	// AgentPresetID identifies the Hecate-owned preset resolved when a
	// project assignment created this task. The effective launch posture is
	// snapshotted onto the task's execution and sandbox fields so later
	// preset edits cannot change retries or resumes.
	AgentPresetID string
	// AgentPresetToolsEnabled snapshots whether the resolved Agent Preset
	// permits tools for this task. nil marks legacy/manual tasks that predate
	// the snapshot and preserves their existing tool behavior; false is an
	// explicit all-tools denial for new preset-backed assignments.
	AgentPresetToolsEnabled *bool `json:",omitempty"`
	// AgentPresetBrowserAllowed snapshots whether the resolved Hecate Agent
	// Preset permits the native, read-only browser evidence tool. nil marks
	// legacy/manual tasks and fails closed: browser evidence is never inferred
	// from SandboxNetwork or a later preset edit.
	AgentPresetBrowserAllowed *bool `json:",omitempty"`
	// AgentPresetBrowserAllowedOrigins is the exact normalized origin allowlist
	// snapshotted with AgentPresetBrowserAllowed. It is deliberately distinct
	// from generic network policy so browser navigation cannot broaden with a
	// task's HTTP settings or a later preset change.
	AgentPresetBrowserAllowedOrigins []string `json:",omitempty"`
	// SystemPrompt is the per-task agent system prompt. When set, it
	// becomes the narrowest layer in the composition: global default →
	// workspace CLAUDE.md/AGENTS.md → this. Concatenated, broadest
	// first. Empty = no per-task add (still honors the broader layers).
	SystemPrompt string
	// WorkspaceSystemPromptPolicy controls the broad workspace prompt
	// compatibility layer. Empty / inherit keeps the default CLAUDE.md /
	// AGENTS.md behavior. Project assignments set exclude so profile
	// context-source policy is the only way to load workspace instruction
	// bodies into their prompt.
	WorkspaceSystemPromptPolicy string
	Repo                        string
	BaseBranch                  string
	WorkspaceMode               string
	// WorkspaceReuse preserves an already-provisioned, runtime-owned
	// workspace when a new task segment must be created. Unlike in_place, the
	// operator-selected source remains isolated; the new task simply continues
	// in the prior managed root instead of cloning it and dropping uncommitted
	// or untracked work. Public task creation does not expose this flag.
	WorkspaceReuse bool
	ExecutionKind  string
	// WorkflowMode and WorkflowVersion identify a bounded Hecate runtime
	// contract. They are blank for the ordinary task path. WorkflowVersion is
	// stored with the task so retries and resumes do not silently adopt future
	// runbook behavior.
	WorkflowMode       WorkflowMode
	WorkflowVersion    string
	ExecutionProfile   string
	OriginKind         string
	OriginID           string
	ShellCommand       string
	GitCommand         string
	WorkingDirectory   string
	FileOperation      string
	FilePath           string
	FileContent        string
	SandboxAllowedRoot string
	SandboxReadOnly    bool
	SandboxNetwork     bool
	// RTKEnabled runs shell/git tool subprocesses through RTK for compact
	// command output. It is persisted on the task so Hecate Chat follow-up
	// runs keep the chat's command-output setting.
	RTKEnabled        bool
	TimeoutMS         int
	Status            string
	Priority          string
	RequestedModel    string
	RequestedProvider string
	BudgetMicrosUSD   int64
	LatestRunID       string
	LastError         string
	CreatedAt         time.Time
	UpdatedAt         time.Time
	StartedAt         time.Time
	FinishedAt        time.Time
	RootTraceID       string
	LatestTraceID     string
	LatestRequestID   string
	// MCPServers configures external MCP servers that an agent_loop run
	// should bring up and expose to the LLM as additional tools. Each
	// entry produces one stdio subprocess (`Command` + `Args`, env
	// merged from `Env`); its tools are registered alongside the
	// built-ins under names of the form `mcp__<server-name>__<tool>`.
	// Empty for non-agent_loop tasks and for agent_loop tasks that
	// don't need external tools.
	MCPServers []MCPServerConfig
}

const (
	WorkspaceSystemPromptInherit = "inherit"
	WorkspaceSystemPromptExclude = "exclude"
)

// MCPEnvEncPrefix is the storage prefix for env values that were
// encrypted at task-creation time with the control-plane AES-GCM
// cipher. The full stored form is "enc:<base64ciphertext>". The MCP
// host factory decrypts these at subprocess spawn time so the plaintext
// token is never written to the task blob.
const MCPEnvEncPrefix = "enc:"

// MCP approval policy values. These control whether the agent loop
// dispatches an MCP tool call from the configured server immediately,
// pauses for operator approval, or refuses to call it at all.
//
//   - MCPApprovalAuto: dispatch immediately. Equivalent to leaving
//     the field empty.
//   - MCPApprovalRequireApproval: pause the agent loop on every call
//     to a tool from this server, emit an approval record, and resume
//     dispatch only after the operator approves.
//   - MCPApprovalBlock: never dispatch; the agent loop returns a tool
//     error to the LLM ("blocked by policy") so the model can pick a
//     different tool without involving the operator.
//
// Per-server granularity is intentional: real configs almost always
// gate or trust an entire server (the github server stays gated, the
// filesystem server stays auto). Per-tool gating is a follow-up.
const (
	MCPApprovalAuto            = "auto"
	MCPApprovalRequireApproval = "require_approval"
	MCPApprovalBlock           = "block"
)

// IsValidMCPApprovalPolicy reports whether v is a recognized policy
// value. Empty string is accepted as the implicit auto default — the
// API layer does not require operators to spell it out.
func IsValidMCPApprovalPolicy(v string) bool {
	switch v {
	case "", MCPApprovalAuto, MCPApprovalRequireApproval, MCPApprovalBlock:
		return true
	default:
		return false
	}
}

// MCPServerConfig describes one external MCP server an agent_loop run
// should connect to. Persisted as part of the task payload (no schema
// migration — the task is stored as a JSON blob in both the sqlite and
// postgres backends).
//
// Env values are stored in one of three forms:
//   - "$VAR_NAME" — resolved from the Hecate process environment at
//     subprocess spawn time; the token itself is never written to the DB.
//   - "enc:<base64>" — encrypted by the API layer with the control-plane
//     AES-GCM key; decrypted at spawn time.
//   - bare literal — stored as-is (acceptable for non-secret values
//     or when no cipher key is configured).
type MCPServerConfig struct {
	// Name is the operator-chosen label used to namespace this server's
	// tools (e.g. "filesystem" → tools become `mcp__filesystem__*`).
	// Must be non-empty and unique within a task; the gateway rejects
	// duplicates at create time.
	Name string
	// Stdio transport (mutually exclusive with URL):
	// Command is the executable that speaks MCP over stdio (e.g. "npx",
	// "uvx", or a path to a binary).
	Command string
	// Args are passed verbatim to the command.
	Args []string
	// Env is merged onto the spawned process's environment. Values may
	// be $VAR_NAME references, enc: ciphertext, or bare literals.
	Env map[string]string
	// HTTP transport (mutually exclusive with Command):
	// URL is the MCP endpoint (e.g. "https://api.example.com/mcp").
	URL string
	// Headers are sent on every HTTP request. Values follow the same
	// $VAR_NAME / enc: / literal rules as Env values.
	Headers map[string]string
	// ApprovalPolicy gates how the agent loop dispatches tool calls
	// from this server. One of "auto" | "require_approval" | "block",
	// or empty (interpreted as auto). See the MCPApproval* constants
	// above for the contract. Per-server, not per-tool.
	ApprovalPolicy string
}

type TaskRun struct {
	ID     string
	TaskID string
	// ProjectID, WorkItemID, and AssignmentID snapshot the project
	// boundary from the parent task at run creation time. They let
	// run streams, traces, and retained run records remain project-aware
	// even when the parent task is not loaded by the caller.
	ProjectID    string
	WorkItemID   string
	AssignmentID string
	Number       int
	Status       string
	Orchestrator string
	// WorkflowMode and WorkflowVersion snapshot the parent task's runtime
	// contract at run creation. They keep retained runs self-describing when
	// task defaults evolve later.
	WorkflowMode       WorkflowMode
	WorkflowVersion    string
	Model              string
	Provider           string
	ProviderKind       string
	WorkspaceID        string
	WorkspacePath      string
	StepCount          int
	ModelCallCount     int
	ApprovalCount      int
	ArtifactCount      int
	TotalCostMicrosUSD int64
	// PriorCostMicrosUSD is the cumulative LLM spend of every prior
	// run in this run's resume chain (excluding this run itself).
	// Fresh runs are zero; resumed/retry-from-model-call runs inherit the
	// source's PriorCost + Total. The per-task cost ceiling check
	// uses (PriorCost + this run's running spend) so a chain of
	// resumes can't escape the ceiling — without this a $5 ceiling
	// could be exceeded by the operator simply re-resuming N times.
	PriorCostMicrosUSD int64
	LastError          string
	StartedAt          time.Time
	FinishedAt         time.Time
	RequestID          string
	TraceID            string
	RootSpanID         string
	OtelStatusCode     string
	OtelStatusMessage  string
	ContextPacket      json.RawMessage
	// InputRef is an opaque, application-owned reference to rich input that
	// must be hydrated immediately before execution. The task runtime persists
	// the reference, never the referenced binary body. Same-input resume and
	// retry runs inherit it; chat continuations replace or clear it. Public
	// task-run renderers intentionally omit it.
	InputRef string
	// InputProviderInstance is the opaque provider instance admitted for
	// InputRef. Admission can resolve it before the final gateway route, so it
	// protects the provider generation while policy and routing resolve the
	// concrete model. It contains no credential or endpoint data and public
	// task-run renderers intentionally omit it.
	InputProviderInstance ProviderInstanceIdentity
	// InputProviderDispatchRecorded marks the final gateway route as durable
	// immediately before provider I/O. Unlike InputProviderInstance, it is not
	// set during admission. Once true, the provider, kind, model, and instance
	// are immutable for this rich input across recovery and same-input retries.
	// Public task-run renderers intentionally omit it.
	InputProviderDispatchRecorded bool
	// InputProviderDisclosedInstance records the provider instance that actually
	// received InputRef. It remains empty when admission failed before a
	// provider dispatch, so transcript history cannot mistake an admission
	// fence for a disclosure boundary. Public task-run renderers intentionally
	// omit it.
	InputProviderDisclosedInstance ProviderInstanceIdentity
}

type TaskStep struct {
	ID            string
	TaskID        string
	RunID         string
	ParentStepID  string
	Index         int
	Kind          string
	Title         string
	Status        string
	Phase         string
	Result        string
	ToolName      string
	Input         map[string]any
	OutputSummary map[string]any
	ExitCode      int
	Error         string
	ErrorKind     string
	ApprovalID    string
	StartedAt     time.Time
	FinishedAt    time.Time
	RequestID     string
	TraceID       string
	SpanID        string
	ParentSpanID  string
}

type TaskApproval struct {
	ID             string
	TaskID         string
	RunID          string
	StepID         string
	Kind           string
	Status         string
	Reason         string
	RequestedBy    string
	ResolvedBy     string
	ResolutionNote string
	CreatedAt      time.Time
	ResolvedAt     time.Time
	RequestID      string
	TraceID        string
	SpanID         string
}

type TaskArtifact struct {
	ID          string
	TaskID      string
	RunID       string
	StepID      string
	Kind        string
	Name        string
	Description string
	MimeType    string
	StorageKind string
	Path        string
	ContentText string
	ObjectRef   string
	SizeBytes   int64
	SHA256      string
	Status      string
	CreatedAt   time.Time
	RequestID   string
	TraceID     string
	SpanID      string
}

type TaskRunEvent struct {
	ID        string
	TaskID    string
	RunID     string
	Sequence  int64
	EventType string
	Data      map[string]any
	CreatedAt time.Time
	RequestID string
	TraceID   string
}

func IsTerminalTaskRunStatus(status string) bool {
	switch status {
	case "completed", "failed", "cancelled":
		return true
	default:
		return false
	}
}

File: ui/src/types/task.ts
export type TaskRecord = {
  id: string;
  title: string;
  prompt: string;
  project_id?: string;
  work_item_id?: string;
  assignment_id?: string;
  // Per-task agent_loop system prompt — narrowest layer in the
  // composition (global → workspace CLAUDE.md/AGENTS.md → this).
  system_prompt?: string;
  repo?: string;
  base_branch?: string;
  workspace_mode?: string;
  execution_kind?: string;
  execution_profile?: string;
  agent_preset_id?: string;
  agent_preset_tools_enabled?: boolean;
  origin_kind?: string;
  origin_id?: string;
  shell_command?: string;
  git_command?: string;
  working_directory?: string;
  file_operation?: string;
  file_path?: string;
  file_content?: string;
  sandbox_allowed_root?: string;
  sandbox_read_only?: boolean;
  sandbox_network?: boolean;
  timeout_ms?: number;
  status: string;
  priority?: string;
  requested_model?: string;
  requested_provider?: string;
  budget_micros_usd?: number;
  latest_run_id?: string;
  // What the most recent run actually used after routing (may
  // differ from requested_* when the operator picked "auto" or
  // the router substituted). Surfaced in the task list.
  latest_model?: string;
  latest_provider?: string;
  pending_approval_count?: number;
  step_count?: number;
  artifact_count?: number;
  last_error?: string;
  created_at?: string;
  updated_at?: string;
  started_at?: string;
  finished_at?: string;
  root_trace_id?: string;
  latest_trace_id?: string;
  latest_request_id?: string;
  // MCPServers echoes the configured external MCP servers (if
  // any). Used by the task list to show an "MCP × N" chip and the
  // task detail to render the per-server configuration. Mirrors
  // the wire shape — see api.MCPServerConfigItem on the gateway
  // side. Secret values (env, headers) come back redacted unless
  // they're $VAR_NAME references; approval_policy and url/command
  // are surfaced verbatim.
  mcp_servers?: Array<{
    name: string;
    command?: string;
    args?: string[];
    env?: Record<string, string>;
    url?: string;
    headers?: Record<string, string>;
    approval_policy?: string;
  }>;
};

export type TasksResponse = {
  object: string;
  data: TaskRecord[];
};

export type TaskResponse = {
  object: string;
  data: TaskRecord;
};

export type TaskRunRecord = {
  id: string;
  task_id: string;
  project_id?: string;
  work_item_id?: string;
  assignment_id?: string;
  number: number;
  status: string;
  orchestrator?: string;
  model?: string;
  provider?: string;
  provider_kind?: string;
  workspace_id?: string;
  workspace_path?: string;
  step_count?: number;
  approval_count?: number;
  artifact_count?: number;
  total_cost_micros_usd?: number;
  // prior_cost_micros_usd is the cumulative LLM spend of earlier
  // runs in this run's resume chain. Cumulative = total + prior;
  // useful when a task has been resumed/retried multiple times.
  prior_cost_micros_usd?: number;
  last_error?: string;
  started_at?: string;
  finished_at?: string;
  request_id?: string;
  trace_id?: string;
  root_span_id?: string;
  otel_status_code?: string;
  otel_status_message?: string;
};

export type TaskRunsResponse = {
  object: string;
  data: TaskRunRecord[];
};

export type TaskRunResponse = {
  object: string;
  data: TaskRunRecord;
};

export type TaskStepRecord = {
  id: string;
  task_id: string;
  run_id: string;
  parent_step_id?: string;
  index: number;
  kind: string;
  title?: string;
  status: string;
  phase?: string;
  result?: string;
  tool_name?: string;
  input?: Record<string, unknown>;
  output_summary?: Record<string, unknown>;
  exit_code?: number;
  error?: string;
  error_kind?: string;
  approval_id?: string;
  started_at?: string;
  finished_at?: string;
  request_id?: string;
  trace_id?: string;
  span_id?: string;
  parent_span_id?: string;
};

export type TaskStepsResponse = {
  object: string;
  data: TaskStepRecord[];
};

export type TaskArtifactRecord = {
  id: string;
  task_id: string;
  run_id: string;
  step_id?: string;
  kind: string;
  name?: string;
  description?: string;
  mime_type?: string;
  storage_kind?: string;
  path?: string;
  content_text?: string;
  object_ref?: string;
  size_bytes?: number;
  sha256?: string;
  status?: string;
  created_at?: string;
  request_id?: string;
  trace_id?: string;
  span_id?: string;
};

export type TaskArtifactsResponse = {
  object: string;
  data: TaskArtifactRecord[];
};

export type TaskArtifactResponse = {
  object: string;
  data: TaskArtifactRecord;
};

export type TaskPatchRecord = {
  artifact: TaskArtifactRecord;
  diff: string;
  status: string;
  path?: string;
  before_existed: boolean;
};

export type TaskPatchResponse = {
  object: string;
  data: TaskPatchRecord;
};

export type TaskApprovalRecord = {
  id: string;
  task_id: string;
  run_id: string;
  step_id?: string;
  kind: string;
  status: string;
  reason?: string;
  requested_by?: string;
  resolved_by?: string;
  resolution_note?: string;
  created_at?: string;
  resolved_at?: string;
  request_id?: string;
  trace_id?: string;
  span_id?: string;
};

export type TaskApprovalsResponse = {
  object: string;
  data: TaskApprovalRecord[];
};

// TaskRunStreamTurnCost mirrors the backend `Turn` block on
// TaskRunStreamEventData. Populated only on snapshots driven by an
// `turn.completed` event, so the UI can render a live per-turn
// cost/tokens summary without subscribing to the public events stream.
export type TaskRunStreamTurnCost = {
  turn_index: number;
  step_id?: string;
  cost_micros_usd: number;
  run_cumulative_cost_micros_usd: number;
  task_cumulative_cost_micros_usd: number;
  tool_calls?: number;
};

export type TaskRunStreamEventData = {
  sequence: number;
  terminal?: boolean;
  event_type?: string;
  run: TaskRunRecord;
  steps?: TaskStepRecord[];
  approvals?: TaskApprovalRecord[];
  artifacts?: TaskArtifactRecord[];
  activity?: TaskActivityRecord[];
  turn?: TaskRunStreamTurnCost;
};

export type TaskActivityRecord = {
  id: string;
  type: string;
  status?: string;
  title?: string;
  step_id?: string;
  artifact_id?: string;
  approval_id?: string;
  tool_name?: string;
  kind?: string;
  path?: string;
  summary?: Record<string, unknown>;
  occurred_at?: string;
  terminal?: boolean;
  needs_action?: boolean;
};

export type TaskRunStreamEventResponse = {
  object: string;
  data: TaskRunStreamEventData;
};

export type TaskRunEventRecord = {
  schema_version: string;
  event_id: string;
  task_id: string;
  run_id: string;
  sequence: number;
  occurred_at: string;
  type: string;
  data: Record<string, unknown>;
};

export type TaskRunEventsResponse = {
  object: string;
  data: TaskRunEventRecord[];
};
